python-package/lightgbm/sklearn.py

```diff
@@ -25,32 +25,50 @@
 ]
 
 _LGBM_ScikitCustomObjectiveFunction = Union[
+    # f(labels, preds)
     Callable[
-        [np.ndarray, np.ndarray],
+        [Optional[np.ndarray], np.ndarray],
```

Comment:
### why add all these `Optional`s?

`mypy` is struggling with the facts that `Dataset.get_field()` can return `None`.
* https://github.com/microsoft/LightGBM/blob/2fe2bf0675ef127a29f9bc49511fc0099cf3e140/python-package/lightgbm/basic.py#L2704
* https://github.com/microsoft/LightGBM/blob/2fe2bf0675ef127a29f9bc49511fc0099cf3e140/python-package/lightgbm/sklearn.py#L127
* https://github.com/microsoft/LightGBM/blob/2fe2bf0675ef127a29f9bc49511fc0099cf3e140/python-package/lightgbm/basic.py#L2393

This PR proposes updating the type hints for custom metric and objective functions to match that behavior.

I intentionally chose not to update the user-facing docs about custom metric and objective functions to reflect that the `label`, `group`, and `weights` passed to these functions can technically be `None`... in almost all situations, they should be non-`None`. I don't think complicating the docs is worth it.